pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,315 @@
+	LastError         error
+}
+
+func (c *Command) Reason() string {
+	return fmt.Sprintf("%s/%s", c.Method,
+		lo.Ternary(len(c.ReplacementKeys) > 0, "replace", "delete"))
+}
+
+// NodeClaimKey wraps a nodeclaim.Key with an initialized field to save on readiness checks and identify
+// when a nodeclaim is first initialized for metrics and events.
+type NodeClaimKey struct {
+	nodeclaim.Key
+	// Use a bool track if a node has already been initialized so we can fire metrics for intialization once.
+	// This intentionally does not capture nodes that go initialized then go NotReady after as other pods can
+	// schedule to this node as well.
+	Initialized bool
+}
+
+type UnrecoverableError struct {
+	error
+}
+
+func NewUnrecoverableError(err error) *UnrecoverableError {
+	return &UnrecoverableError{error: err}
+}
+
+func IsUnrecoverableError(err error) bool {
+	if err == nil {
+		return false
+	}
+	var unrecoverableError *UnrecoverableError
+	return errors.As(err, &unrecoverableError)
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	// providerID -> command, maps a candidate to its command. Each command has a list of candidates that can be used
+	// to map to itself.
+	ProviderIDToCommand map[string]*Command
+	mu                  sync.RWMutex
+	kubeClient          client.Client
+	recorder            events.Recorder
+	cluster             *state.Cluster
+	clock               clock.Clock
+	provisioner         *provisioning.Provisioner
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate disruption commands
+func NewQueue(kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock,
+	provisioner *provisioning.Provisioner) *Queue {
+	queue := &Queue{
+		RateLimitingInterface: workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		ProviderIDToCommand:   map[string]*Command{},
+		kubeClient:            kubeClient,
+		recorder:              recorder,
+		cluster:               cluster,
+		clock:                 clock,
+		provisioner:           provisioner,
+	}
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, timeAdded time.Time, method string, consolidationType string) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates:        candidates,
+		Method:            method,
+		ConsolidationType: consolidationType,
+		TimeAdded:         timeAdded,
+	}
+}
+
+func (q *Queue) Name() string {
+	return "disruption.queue"
+}
+
+func (q *Queue) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+func (q *Queue) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
```

Comment:
Is this queue just a *Controller? Why create a new concept?

Reply:
To add to this (I don't think that we should do this now) but it might be worth considering if there's a way to define a `source.Source` in controller runtime that we can use to leverage their queue implementation natively rather than having to build our own implementation here. The only downside of that is that we might lose the queue depth metric with that implementation.

Reply:
Controller runtime provides a workqueue depth metric, IIUC.

Reply:
I don't believe that they do. There's not really a clean way to do it based on how the `DelegatingQueue` works from what I understand since the items aren't actually put into the queue until after their requeue interval has finished.